Scripts/Halls of Origination: Scripts, rewrites, corrections

Mihc3 opened the pull request:
Almost fully blizzlike as always. Source:
15595_2012-08-22_22-02-32_looking_for_dungeon_delete_item_set.pkt and
retail experience.

<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**
- Update instance model and boss scripts
- Some corrections on existing encounters

New stuff:
- Isiset boss scripted
- Some trash SAI and waypoints
- The Maker's Lift Controller scripted

**Target branch(es):** 3.3.5/master
- master

**Tests performed:** (Does it build, tested in-game, etc.)
- [x] Build
- [ ] Test

**Known issues and TODO list:** (add/remove lines as needed)
- [ ] Spatial Flux (39612) won't enter combat on second aggro from creature group (but it should).
    - Links: [cpp](https://github.com/Mihapro/TrinityCore/blob/23ec8d7a0271baeb58c0f35e7772c6769e996da7/src/server/scripts/Kalimdor/HallsOfOrigination/halls_of_origination.cpp#L52), [sql](https://github.com/Mihapro/TrinityCore/blob/23ec8d7a0271baeb58c0f35e7772c6769e996da7/sql/updates/world/master/2017_12_05_20756_world.sql#L10)
- [ ] Target type 112 (TARGET_UNK_112) not scripted.
    - Only used by one spell: 89549/Rune of Healing (trash mob ability).
- Stopped elevators do not spawn on first floor (frame 0 is not registered as a stopFrame), but on second floor (which takes index 0 instead).

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Commit:
Scripts/Halls of Origination: scripts, rewrites, corrections

Almost fully blizzlike as always. Source:
15595_2012-08-22_22-02-32_looking_for_dungeon_delete_item_set.pkt and
retail experience.

joschiwald commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_anraphet.cpp:
```diff
@@ -327,24 +336,37 @@ class npc_brann_bronzebeard_anraphet : public CreatureScript
+
+            player->PrepareGossipMenu(creature, GOSSIP_MENU_WE_ARE_READY, true);
+            player->SendPreparedGossip(creature);
+            return true;
```
this is doable with conditions

Mihapro replied:
I see. Conditions are already there... https://pastebin.com/nWdeR2fL
But creature_template only has slot for one menu id, how does it find others?

tkrokli replied:
`creature_template` only lists the default `gossip_,menu_id` for a creature. To access gossip menus or gossip menu options, link them via the `gossip_menu` table and use conditions for when they are shown to the player.

tkrokli replied:
if you need a gossip menu option to open a new gossip menu, it is done via the `gossip_menu_option`.`action_menu_id` field.

Mihapro replied:
npc must show another ```gossip_menu``` on hello, and there is no link between second menu and ```creature_template``` :/ i guess i will just change ```creature_template```.```gossip_menu_id``` manually

tkrokli replied:
Which creature ID (`creature_template`.`entry`) and `creature_template`.`gossip_menu_id` are the original ones in this case?

Mihapro replied:
Brann 39908
Menu that starts the encounter 11339

tkrokli replied:
Well, it can possibly be set in `conditions` by using `CONDITION_SOURCE_TYPE_GOSSIP_MENU` 14, `SourceGroup` (gossip menu entry) 11339, SourceEntry (gossip_menu.text_id) 15794 and with `ConditionTypeOrReference` `CONDITION_INSTANCE_INFO` 13 -- see https://trinitycore.atlassian.net/wiki/spaces/tc/pages/2130002/conditions#conditions-ConditionTypeOrReference for more info. Maybe @Killyana can suggest some more to help you along the way here.

Mihapro replied:
[198_misc.sql#L2](https://github.com/TrinityCore/TrinityCore/blob/a249d86b00eb0854cb925f8bea76ddbb364208df/sql/old/4.3.4/TDB1_to_TDB2_updates/world/198_misc.sql#L2) - here, back in 2014, it was set so it works, but sniffs overwrite values.
Anyway, for now it seems only C++ can handle menu change...

tkrokli replied:
That's odd. I wonder why the `gossip_menu_id` changed, but not the `conditions` table or `gossip_menu` table content for this encounter.

Mihapro replied:
He ends up with different gossip text after you kill Anraphet... different ```gossip_menu_id ```. Probably that's why

joschiwald commented on src/server/scripts/Kalimdor/HallsOfOrigination/instance_halls_of_origination.cpp:
```diff
@@ -207,9 +259,15 @@ class instance_halls_of_origination : public InstanceMapScript
+                    if (GameObject* mirror = instance->GetGameObject(SunMirrorGUID))
+                        mirror->SetGoState(GO_STATE_ACTIVE);
+                    if (GameObject* door = instance->GetGameObject(AnraphetDoorGUID))
+                        door->SetGoState(GO_STATE_ACTIVE);
```
i don't think any gameobject is spawned at this moment

Mihapro replied:
It's just rewritten, should it be moved to OnGameObjectCreate()?

joschiwald replied:
should be better

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_anraphet.cpp:
```diff
@@ -327,24 +337,37 @@ class npc_brann_bronzebeard_anraphet : public CreatureScript
-            npc_brann_bronzebeard_anraphetAI(Creature* creature) : CreatureAI(creature), _currentPoint(0), _instance(creature->GetInstanceScript()) { }
+            npc_brann_bronzebeard_anraphetAI(Creature* creature) : CreatureAI(creature), _currentPoint(0), _instance(creature->GetInstanceScript())
+            {
+                events.ScheduleEvent(EVENT_BRANN_IDLE_EMOTE, 45000);
```
it should be on Reset(), and add std::chrono, like:

events.ScheduleEvent(EVENT_BRANN_IDLE_EMOTE, Seconds(45));

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_anraphet.cpp:
```diff
@@ -379,6 +404,14 @@ class npc_brann_bronzebeard_anraphet : public CreatureScript
                     {
+                        case EVENT_BRANN_IDLE_EMOTE:
+                            me->HandleEmoteCommand(EMOTE_ONESHOT_USE_STANDING);
+                            events.ScheduleEvent(EVENT_BRANN_IDLE_EMOTE, 45000);
```
events.Repeat(Seconds(45));

Mihapro replied:
Also in boss scripts?

Keader replied:
yeap

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+#include "halls_of_origination.h"
+#include "TemporarySummon.h"
+
+enum Spells
```
identation will be nice xD

Mihapro replied:
lazy :(

tkrokli replied:
(already fixed, but does not hide these comments because comment was placed outside of text to be changed)

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+        {
+            switch (summoned->GetEntry())
+            {
+            case NPC_ASTRAL_RAIN:
```
identation

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+            switch (summoned->GetEntry())
+            {
+            case NPC_ASTRAL_RAIN:
+                instance->SendEncounterUnit(ENCOUNTER_FRAME_ENGAGE, summoned, 1);
```
this creatures has core scripts? it can be handled on Reset() or EnterCombat() of this creatures no?

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+            {
+                switch (eventId)
+                {
+                case EVENT_START_ATTACK:
```
identation xd

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+                    DoStartMovement(me->GetVictim());
+                    break;
+                case EVENT_ASTRAL_RAIN:
+                    DoCast(me, SPELL_ASTRAL_RAIN_CONTROLLER);
```
DoCastSelf()

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+                    break;
+                case EVENT_ASTRAL_RAIN:
+                    DoCast(me, SPELL_ASTRAL_RAIN_CONTROLLER);
+                    events.ScheduleEvent(EVENT_ASTRAL_RAIN, urand(15,19)*IN_MILLISECONDS);
```
events.ScheduleEvent(EVENT_ASTRAL_RAIN, Seconds(15), Seconds(19));

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+                }
+                default:
+                    break;
+                }
```
Missing a if (me->HasUnitState(UNIT_STATE_CASTING)) return; here

Mihapro replied:
There is one before event execute switch

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+
+            DoCast(SPELL_ASTRAL_SHIFT);
+            DoCast(SPELL_MIRROR_IMAGE_STARRY_SKY_SPAWNER);
+            DoCast(SPELL_MIRROR_IMAGE_SPAWNER);
```
has no category lock in this 3 spells? i mean, he can cast this 3 spells together without trigger flag?

Mihapro replied:
In sniffs, SPELL_MIRROR_IMAGE_SPAWNER is the last spell he casts, so it's not triggered, it triggers 3 other spells. Same goes for SPELL_MIRROR_IMAGE_STARRY_SKY_SPAWNER.

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1071 @@
+
+        void UpdateAI(uint32 diff) override
+        {
+            events.Update(diff);
```
No victim check? creature will never enter in evade :/

Mihapro replied:
It needs to execute some events while reacting passive, I couldn't remember any other way :/

Mihapro replied:
It's a boss summon and despawns anyway

Mihapro replied:
Enters combat rightaway now, so UpdateVictim check is back.

tkrokli commented:
```
src/server/scripts/Kalimdor/HallsOfOrigination/boss_anraphet.cpp:349:31:
 fatal error: use of undeclared identifier 'GOSSIP_MENU_WE_ARE_READY'
                if (menuId == GOSSIP_MENU_WE_ARE_READY)
                              ^
```
Looks like you need to add back the Gossip enum if you want to use the enum name in the script.
```cpp
enum Gossip
{
    GOSSIP_MENU_WE_ARE_READY = 11339,
};
```

tkrokli commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,872 @@
+enum Spells
+{
+    SPELL_SUPERNOVA                         = 74136,
+    SPELL_SUPERNOVA_EFFECT = 74137,
```
BTW, this line "needs more cowbell" ;-) :-P

tkrokli commented on sql/updates/world/master/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,238 @@
+
+-- Isiset Text
+DELETE FROM `creature_text` WHERE `entry` IN (39587);
+INSERT INTO `creature_text` (`entry`, `groupid`, `id`, `text`, `type`, `language`, `probability`, `emote`, `duration`, `sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
```
You need to update this query to match the new `creature_text` template:
`CreatureID, GroupID, ID, Text, Type, Language, Probability, Emote, Duration, Sound`
(last 3 field names are unchanged).

Mihapro replied:
Thanks! 👍

tkrokli commented on sql/updates/world/master/9999_99_99_99_world.sql:
```diff
@@ -34,7 +34,7 @@ INSERT INTO `conditions` (`SourceTypeOrReferenceId`, `SourceGroup`, `SourceEntry
 
 -- Isiset Text
 DELETE FROM `creature_text` WHERE `entry` IN (39587);
```
Oh. I forgot to specify that you need to change this one too,
`entry` `->` `CreatureID`

tkrokli commented:
Give us a hint when you think you are done with your changes, so we can change the label again. :-)

Mihapro commented:
@tkrokli one more heroic-mode ability to script, hopefully today. 😀

Mihapro commented:
There, everything seems to be working - except that everytime one of shared npc scripts (either `npc_hoo_spatial_flux` or `new npc_hoo_energy_flux`) won't load on second npc: "no core script found".

Strange because three mirror npcs share the same script just fine (`npc_isiset_mirror_image`).

tkrokli commented:
Did your latest commit fix the issue you mentioned in your comment https://github.com/TrinityCore/TrinityCore/pull/20756#issuecomment-342038786 ?

Mihapro commented:
Sadly no. I wonder if it has something to do because of script's location - it is not in the same file as Isiset.
`npc_hoo_spatial_flux ` is in `halls_of_origination.cpp`
`boss_isiset` is in `boss_isiset.cpp`
`Isiset->SummonCreature(SpatialFlux ...` It never loads. 😞

tkrokli commented:
Hmm, that's a bit bothersome. Maybe `npc_hoo_spatial_flux` needs some form of linking via `halls_of_origination.h` , unless it can be linked directly somehow, like via `#include`

Mihapro commented:
Now I remembered that there is no summon packet. Could it mean it uses different way to spawn it? Some Mgr or something

tkrokli commented:
OK, you mean something like
```cpp
void OnCreatureCreate(Creature* creature) override
```
or related code? :-)

Mihapro commented:
Meh, I'll separate them and see if it gets better. 😄

Mihapro commented:
It's fine now. 👍 A new [minor bug](https://github.com/TrinityCore/TrinityCore/pull/20756/commits/1e98fe12149e976008326c5b75ed832e512411bd#diff-d682fe09f0c4bfe6a8921095cb869971R49) appears tho 😄

tkrokli commented:
OK, but what about your "**Known issues and TODO list**" points?
Are they correct or do they need to be updated?

Mihapro commented:
Just referencing some issues that would make this instance better, nothing major tho 😃

tkrokli commented:
Very well, but if you list issues as part of the description, it helps if you write something like "(Outside the scope of this PR.)" behind a listed issue, when the listed issue can not be solved within this PR. Otherwise it will look like you still have got things to work no before this PR is finished.

Mihapro commented:
This PR can be pushed. 👍

tkrokli commented:
Even if it is mostly cosmetic, I must admit that I like the look of the SQL formatting in this PR. If I could, I would have printed it out on some kind of material to use as my dinner table mat so I could look at it during most of my meals. 😄

BTW: could you browse through once more and check for any left-over "To-Do" points like this text in the SQL: `-- Script names, heroic entry (to-do)` and update any such text to show the actual state of your code? :-)

Commit:
Rename 9999_99_99_99_world.sql to 2017_12_05_20756_world.sql

- temporary merge preparation

Commit:
Scripts/Halls of Origination: Scripts, rewrites, corrections

tkrokli commented:
I am a bit slow at understanding your intentions when it comes to the "to-do" bits. Does this mean that there will be one or more "to-do" points left in the code even if you said that you consider this PR to be finished from your point of view?+

Mihapro commented:
I just note everything that doesn't work perfectly (2.) or isn't scripted (1.), but those are minor things and depending on how small my 'npc_hoo_spatial_flux' script is, it's probably some general core combat issue (I still like to make notes as comments - like in SQL, you know 😝).

tkrokli commented:
Hmm, OK. Ideally, we should avoid leaving extra comments in PR code about unsolved issues, but if it can't be avoided I suppose it will have to stay (unless someone else has suggestions about how to fix remaining issues before merging the PR content). I will leave the current PR status as it is, until one of the main TC developers says if it is good enough to merge as it is or if it needs to be improved first.

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_anraphet.cpp:
```diff
@@ -50,27 +51,36 @@ enum Texts
 
+enum Gossip
+{
+    GOSSIP_MENU_WE_ARE_READY          = 11339,
```
comma xD

tkrokli replied:
cosmetic/codestyle, but yeah :-P

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_anraphet.cpp:
```diff
@@ -331,20 +341,24 @@ class npc_brann_bronzebeard_anraphet : public CreatureScript
                     me->SetWalk(true);
-                    Talk(BRANN_SAY_DOOR_INTRO);
-                    events.ScheduleEvent(EVENT_BRANN_UNLOCK_DOOR, 7500);
+                    events.ScheduleEvent(EVENT_BRANN_START_INTRO, 1000);
```
missing chrono

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_anraphet.cpp:
```diff
@@ -355,12 +369,14 @@ class npc_brann_bronzebeard_anraphet : public CreatureScript
+                        else
                         {
+                            events.ScheduleEvent(EVENT_BRANN_4_ELEMENTAL_DEAD_DOOR, Seconds(9));
+                            events.ScheduleEvent(EVENT_BRANN_4_ELEMENTAL_DEAD_SAY, 13500);
```
chrono

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_anraphet.cpp:
```diff
@@ -379,6 +395,14 @@ class npc_brann_bronzebeard_anraphet : public CreatureScript
+                            break;
+                        case EVENT_BRANN_START_INTRO:
+                            Talk(BRANN_SAY_DOOR_INTRO);
+                            events.ScheduleEvent(EVENT_BRANN_UNLOCK_DOOR, 7500);
```
chrono

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+    SPELL_CALL_OF_SKY                       = 90750, // Summons Spatial Flux
+    SPELL_SPAWN_ENERGY_FLUX_ISISET          = 90735, // Makes random player cast Summon Energy Flux
+    SPELL_ENERGY_FLUX_BEAM_ISISET           = 90741, // Makes nearby Spatial Flux cast visual beam
+    SPELL_ENERGY_FLUX_PERIODIC              = 74044,
```
comma

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+    NPC_ASTRAL_RAIN                         = 39720, // N
+    NPC_CELESTIAL_CALL                      = 39721, // E
+    NPC_VEIL_OF_SKY                         = 39722, // W
+    NPC_ISISET_SPATIAL_FLUX                 = 48707, // heroic only
```
comma

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+    SAY_SUPERNOVA                           = 1,
+    SAY_SUPERNOVA_WARNING                   = 2,
+    SAY_PLAYER_KILL                         = 3,
+    SAY_DEATH                               = 4,
```
comma

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+
+enum Events
+{
+    EVENT_NONE,
```
i got some weird behaviors before, with enums not initialized.
i always define the first value xD

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+            me->SetReactState(REACT_AGGRESSIVE);
+
+            transitionPhase = false;
+            instance->SetData(DATA_ISISET_PHASE, 1);
```
magic numbers
and unneded space above

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+            // Handle despawning
+            DummyEntryCheckPredicate pred;
+            summons.DoAction(ACTION_IMAGES_SET_PASSIVE, pred);
+            events.ScheduleEvent(EVENT_DESPAWN_IMAGES, 2400);
```
chrono

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+
+        void JustDied(Unit* /*killer*/) override
+        {
+            //instance->SendEncounterUnit(ENCOUNTER_FRAME_SET_COMBAT_RES_LIMIT, ?me?, 1);
```
remove?

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+                    }
+                    default:
+                        break;
+                }
```
Missing a
```
if (me->HasUnitState(UNIT_STATE_CASTING))
    return;
```

here

Mihapro replied:
Added 👍

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+        void RescheduleEvents()
+        {
+            events.Reset();
+            if (instance->GetData(DATA_ISISET_ASTRAL_RAIN_ALIVE) == 1)
```
== 1 not needed :)

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+            events.ScheduleEvent(EVENT_IMAGES_ATTACK, Seconds(2));
+        }
+
+        bool transitionPhase;
```
trinity standard: private variables should start with a ```_``` , e.g. ```bool _transitionPhase; ```

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+            me->SetInCombatWithZone();
+            events.ScheduleEvent(EVENT_ORB_ADD_VISUAL, 1200);
+            events.ScheduleEvent(EVENT_ORB_SET_AGGRESSIVE, 2400);
+            events.ScheduleEvent(EVENT_ORB_ARCANE_BARRAGE, Seconds(urand(4,6)));
```
urand is not needed anymore (and missing chrono above)

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+                    default:
+                        break;
+                }
+            }
```
if (me->HasUnitState(UNIT_STATE_CASTING))
                return;

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+        }
+
+    private:
+        EventMap events;
```
not used

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+    {
+        npc_isiset_mirror_imageAI(Creature* creature) : ScriptedAI(creature), _instance(creature->GetInstanceScript())
+        {
+            me->SetReactState(REACT_PASSIVE);
```
Move everything here to Reset() hook

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+                case ACTION_IMAGES_SET_AGGRESSIVE:
+                    me->SetReactState(REACT_AGGRESSIVE);
+                    me->SetInCombatWithZone();
+                    AttackStart(SelectTarget(SELECT_TARGET_RANDOM));
```
Check if generic AttackStart has protection to null.
Because SelectTarget(SELECT_TARGET_RANDOM) can return null xd

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+                        break;
+                    case EVENT_IMAGE_VEIL_OF_SKY_ABILITY:
+                        DoCastSelf(SPELL_VEIL_OF_SKY_1); // Veil of Sky npc does not cast controller spell!
+                        //_events.Repeat(Seconds(20), Seconds(25)); // It does probably not repeat.
```
so, remove?

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+                switch (eventId)
+                {
+                    case EVENT_DUMMY_NUKE:
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 100, true))
```
distance is a float, not a int (missing .0f)

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+                    default:
+                        break;
+                }
+            }
```
missing a unit state cast check

Mihapro replied:
Added 👍

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+
+        bool Validate(SpellInfo const* /*spellInfo*/) override
+        {
+            return ValidateSpellInfo({ SPELL_ASTRAL_RAIN_1, SPELL_ASTRAL_RAIN_2, SPELL_ASTRAL_RAIN_3 });
```
1 spell per line

tkrokli replied:
Like this?
```cpp
        bool Validate(SpellInfo const* /*spellInfo*/) override
        {
            return ValidateSpellInfo(
            {
                SPELL_ASTRAL_RAIN_1,
                SPELL_ASTRAL_RAIN_2,
                SPELL_ASTRAL_RAIN_3
            });
        }
```

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+        void HandleDummy(SpellEffIndex /*effIndex*/)
+        {
+            InstanceScript* instance = GetCaster()->GetInstanceScript();
+            if (!instance || instance->GetData(DATA_ISISET_ASTRAL_RAIN_ALIVE) == 0)
```
!instance->GetData(DATA_ISISET_ASTRAL_RAIN_ALIVE) instead

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+
+            uint32 phase = instance->GetData(DATA_ISISET_PHASE);
+            if (phase == 1)
+                GetCaster()->CastSpell(GetCaster(), SPELL_VEIL_OF_SKY_1, false);
```
false is default value, can be omitted

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+            if (!instance)
+                return;
+
+            if (instance->GetData(DATA_ISISET_CELESTIAL_CALL_ALIVE) == 1)
```
if (instance->GetData(DATA_ISISET_CELESTIAL_CALL_ALIVE))

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+        void OnMissleHit(SpellEffIndex /*effIndex*/)
+        {
+            if (InstanceScript* instance = GetCaster()->GetInstanceScript())
+                if (Creature* Isiset = ObjectAccessor::GetCreature(*GetCaster(), instance->GetGuidData(DATA_ISISET)))
```
isiset

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+            // Remove tank
+            if (InstanceScript* instance = GetCaster()->GetInstanceScript())
+                if (Creature* Isiset = ObjectAccessor::GetCreature(*GetCaster(), instance->GetGuidData(DATA_ISISET)))
+                    if (WorldObject* tank = Isiset->AI()->SelectTarget(SELECT_TARGET_TOPAGGRO))
```
isiset

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1075 @@
+        npc_astral_shift_explosion_visualAI(Creature* creature) : ScriptedAI(creature)
+        {
+            DoCastSelf(SPELL_ASTRAL_SHIFT_EXPLOSION_VISUAL);
+            me->DespawnOrUnsummon(21300);
```
missing chrono

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/halls_of_origination.cpp:
```diff
@@ -0,0 +1,251 @@
+                switch (eventId)
+                {
+                    case EVENT_FOLLOW_SUMMONER:
+                        me->GetMotionMaster()->MovePoint(0, target->GetPosition(), true);
```
Why not MoveChase?

Mihapro replied:
Need to follow with zero distance, I don't think any other movement worked.

Mihapro replied:
https://hastebin.com/aficezomib.vbs

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/halls_of_origination.cpp:
```diff
@@ -0,0 +1,251 @@
+        {
+            if (GetTarget()->GetPowerPct(POWER_ENERGY) == 100.0f)
+            {
+                GetTarget()->CastSpell((Unit*)nullptr, SPELL_ARCANE_BURST, false);
```
This cast is needed (Unit*) nullptr?
idk if already ported to master (is not needed anymore in 3.3.5)
and remove this false :)

Keader commented on src/server/scripts/Kalimdor/HallsOfOrigination/instance_halls_of_origination.cpp:
```diff
@@ -190,13 +227,32 @@ class instance_halls_of_origination : public InstanceMapScript
+                    {
+                        CreatureGroup* group = creature->GetFormation();
+                        // Check if whole formation is dead, then despawn its leader
+                        for (GuidVector::const_iterator itr = isisetTrashGUIDs.begin(); itr != isisetTrashGUIDs.end(); ++itr)
```
why?
You can use a foreach here o.o
for (ObjectGuid guid : isisetTrashGUIDs)

Mihapro commented:
Can’t get elevator to work using setTransportState :( it should be stopped but still probably move/teleport if stopFrame is set

Warpten commented:
The elevator in HoO is weird, it has two stop frames, and should also rotate while ascending/descending (doing a 180 from ground 0 to 1, 1 to 2, etc). This is handled (at least on 4.3.4) by the transport position element of regular player movement (not implemented on TC, afaik, we need orientation interpolation between frames). However, because of how partially we are supporting non-`MO_TRANSPORT`s, the server does not consider elevators to be transports - they are handled as regular gameobjects. Their positions desynchronizes between clients along with a hefty bag of issues (due to grid and something else). Also it is worth noting that their handling has considerably changed by both the client and the server between wrath and WOD/legion, and cata is just no different. I have pending work on this but it is just a pure shithole that requires a lot of work to be put in so that we can have proper support of the most basic things like seeing people correctly go up and down on an elevator like the one in BoT.

Mihapro commented:
I see, I will investigate more later. I saw frames/stop frames are already set in dbc, couldn’t we just read those instead of db data1, data6 etc?

Mihapro commented:
@Shauren gameobjects wont rotate around x and y :( everytime it just makes some x and y offset to position instead of rotation, and even resets orientation to 0. Do we have 7.x sniffs of any go that is rotated like that?

tkrokli commented on sql/updates/world/master/2017_12_05_20756_world.sql:
```diff
@@ -0,0 +1,262 @@
+(12512, 17600, 0);
+
+DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_anhuur_activate_beacons'; -- rewritten script: spell_anhuur_activate_beacons -> spell_anhuur_handle_beacons
+DELETE FROM `spell_script_names` WHERE `ScriptName` IN ('spell_anhuur_handle_beacons,', 'spell_isiset_veil_of_sky', 'spell_isiset_supernova_filter', 'spell_isiset_mirror_image_starry_sky_spawner', 'spell_isiset_mirror_image_spawner', 'spell_isiset_image_explosion', 'spell_isiset_astral_rain_controller', 'spell_isiset_mana_shield_controller', 'spell_isiset_astral_familiar_controller', 'spell_isise ...
```
```sql
[...] WHERE `ScriptName` IN ('spell_anhuur_handle_beacons,', '
```
redundant comma at the end of the spell name, before the tick.
BTW, is there a (cosmetic/codestyle) standard for sorting the list below?

Mihapro replied:
No idea, i will reorganize into titled sections for easier reading

tkrokli commented on sql/updates/world/master/2017_12_05_20756_world.sql:
```diff
@@ -0,0 +1,283 @@
+
+-- Boss: Isiset -- (to-do: heroic entry)
+-- heroic entry
+INSERT INTO `creature_template` (`entry`, `difficulty_entry_1`, `difficulty_entry_2`, `difficulty_entry_3`, `KillCredit1`, `KillCredit2`, `modelid1`, `modelid2`, `modelid3`, `modelid4`, `name`, `femaleName`, `subname`, `IconName`, `gossip_menu_id`, `minlevel`, `maxlevel`, `HealthScalingExpansion`, `RequiredExpansion`, `VignetteID`, `faction`, `npcflag`, `speed_walk`, `speed_run`, `scale`, `rank`, ...
```
This line needs an opening statement like
```sql
DELETE FROM `creature_template` WHERE `entry`=39586;
```
, doesn't it? (BTW, looks like a github bug above, not highlighting the `INSERT INTO` syntax)

Mihapro replied:
or maybe intended to be black? not sure, i dont comment much :D ty, ill add that delete

Commit:
Brann stops at first point

Commit:
Sniffed corrections: inhabitType, Brann's speed

Commit:
Misc

Commit:
Misc

* correct rotation for go Glass Star 2
* wrong orientation for Anhuur, too old sniffs

Commit:
Misc

* spawned missing Temple Fireshaper
* exit shield phase when Hymn channeling finishes

Mihapro commented:
Self reminder for tomorrow: debug crash after ExitShieldPhase() if Hymn channel expires

Commit:
Fix crash, clean up

Commit:
Elemental scripts, movement for some troggs, ...

Commit:
Troggs, why did it have to be Troggs!

complete respawn, waypoints, to-do: random movement

Commit:
Misc

Commit:
HoO misc

* spawn Transit Devices after bosses die
* wardens now randomly switch between platforms
* some corrections

Commit:
Misc

* Achievement credit spell added to dbc
* Fix wardens not updating orientation visually
* Despawn troggs if Anraphet is activated

Commit:
Misc 2

Commit:
Transit Devices

Commit:
Anhuur

* (Re)schedule events after phase change (EventMap deletes event
executions with different phase - EventMap.cpp:71)

Commit:
Earthrager Path rewritten

Commit:
Fix build

Commit:
missing override

Commit:
More trash SAI, Brann's AreaTrigger, ...

* Deactivate beacons if Anhuur finishes channeling Hymn.

Commit:
fix sql

Warpten commented:
@Mihapro I don't follow

Commit:
Update to-do

Commit:
areatrigger sql

Commit:
Anhuur: some visuals cast with more delay

* Remove wrong spellfix, add another to replace spellscript
* Fix crash: check that go is not null before calling
HandleTransitDevice

Commit:
Earthrager Ptah

* stop channel if its time for sandstorm
* attempt to fix trash submerge/emerge

Commit:
Misc

AreaTrigger ID correction, trash heroic entries and more

tkrokli commented on sql/updates/world/master/9999_99_99_01_world.sql:
```diff
@@ -0,0 +1,835 @@
+UPDATE `creature_template` SET `ScriptName` = 'npc_hoo_aggro_stalker' WHERE `entry` = 40790;
+UPDATE `creature_template` SET `ScriptName` = 'npc_hoo_spatial_flux', `unit_flags` = 33554496 WHERE `entry` = 39612;
+UPDATE `creature_template` SET `ScriptName` = 'npc_hoo_energy_flux', `unit_flags` = 33554496, `speed_walk` = 5.5/2.5, `speed_run` = 5.5/2.5 WHERE `entry` = 44015;
+DELETE FROM `spell_script_names` WHERE `ScriptName` IN ('spell_hoo_emerge', 'spell_hoo_energy_flux_target_selector', 'spell_hoo_arcane_energy_check');
```
cosmetic/codestyle: add a blank line between line 15 and 16 (different tables + query types)

Commit:
Misc

Commit:
Misc

Merge troggs sql for creature table, attempt to fix achievements, fix
crash/missing break, Ptah wont enter disperse if he is about to die, ...

Commit:
HoO

* Fix crash, safer handling of warden spell positions
* Achievement: Instance script -> script

Commit:
Less code (y)

Commit:
Achievement dbc fix, comment correction

Ovahlord commented:
The elevator issue is caused by an known issue that elevators are not being stopped properly. 
https://github.com/TrinityCore/TrinityCore/issues/14704
And I cannot remember any rotation of the HoO elevator at all o.o

Mihapro commented:
Yeah it doesn’t. We continued discussion on irc :P atm it doesn’t stop at frame 0 because we don’t add is as first floor before adding second, third... (gameobject.cpp i think)

Commit:
latest changes

Commit:
Corrections

Commit:
activate correct laserbeam just like lightmachine

Commit:
Ammunae, Setesh, trash scripted

Does not build yet. Rajh to-do.

Aokromes commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_ammunae.cpp:
```diff
@@ -0,0 +1,246 @@
+/*
+* Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
2018 xd

Aokromes commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_setesh.cpp:
```diff
@@ -0,0 +1,512 @@
+/*
+* Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
2018

Aokromes commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_setesh.cpp:
```diff
@@ -0,0 +1,512 @@
+    new npc_setesh_chaos_seed();
+    new npc_setesh_chaos_portal();
+    new spell_setesh_chaos_blast();
+}
```
missing newline

Mihapro replied:
added

Aokromes commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_ammunae.cpp:
```diff
@@ -0,0 +1,246 @@
+{
+    new boss_ammunae();
+    new spell_ammunae_consume_life_energy();
+}
```
missing newline

Mihapro replied:
added

tkrokli commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1103 @@
+/*
+ * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
`2017 -> 2018`

tkrokli commented on src/server/scripts/Kalimdor/HallsOfOrigination/halls_of_origination.cpp:
```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
2018

Ovahlord reviewed (commented):
you can use my rajh script from 4.3.4 :P

Commit:
Fix build

* correct prediction type for SelectTarget
* missing EnterEvadeMode argument
* http -> https, interruptable typo

Commit:
Misc

* Earthrager Ptah creature text
* missing overrides, clean-up

Commit:
Travis fix

tkrokli commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1103 @@
+
+enum Events
+{
+    EVENT_SUPERNOVA,
```
enum = 0 here, does this event really work?
IIRC event numbering should start from 1, not 0.

Mihapro replied:
yeah 0 doesn't work, missing = 1

Mihapro replied:
Funny thing it did work as 0... Supernova timer worked, it wasn't spammed 🤔

tkrokli replied:
Interesting. I guess the value can't be taken for granted, then.

Mihapro replied:
Seems like (int)0 enters default case in switch and (uint32)0 correctly fires enum EVENT_SUPERNOVA = 0 ...

It is strange, but I was testing/player Isiset for weeks and it really worked 😄

Commit:
Crash fix

Commit:
Misc

* Isiset: event indexing starts at 1
* Setesh: timer corrections
* Clean-up

Commit:
Travis fix

Commit:
Ammunae timer corrections

Commit:
Misc

* Chaos Portal: wave events instead of add type events
* cosmetic: Earthrager case before Vault of Lights

tkrokli commented on src/server/scripts/Kalimdor/HallsOfOrigination/halls_of_origination.cpp:
```diff
@@ -0,0 +1,414 @@
+enum Events
+{
+    // Spatial Flux
+    EVENT_DUMMY_NUKE,
```
Event enum starts on 0 instead of 1 here.
Does it work as intended?

Mihapro replied:
Dummy Nuke is unscripted, unknown purpose... I will still update it to 1 👍

tkrokli replied:
Saw your reply to the previous one just now.
Feel free to ignore my previous comment.

joschiwald commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,1103 @@
+
+    private:
+        EventMap events;
+        Unit* target;
```
don't save pointers

joschiwald commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_setesh.cpp:
```diff
@@ -0,0 +1,531 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
bad indentation

joschiwald commented on src/server/scripts/Kalimdor/HallsOfOrigination/halls_of_origination.cpp:
```diff
@@ -0,0 +1,414 @@
+
+        void HandleScriptEffect(SpellEffIndex effIndex)
+        {
+            uint32 spellId = GetSpellInfo()->GetEffect(effIndex)->BasePoints;
```
GetEffectValue()

Commit:
Misc sql

Commit:
Some npcs to SAI, cosmetics

Shrink spell to-do, but customs spells wont load anyway :(

Aokromes commented on sql/updates/world/master/9999_99_99_01_world.sql:
```diff
@@ -35,8 +35,8 @@ INSERT INTO `spelleffect_dbc` (`Id`, `Effect`, `EffectValueMultiplier`, `EffectA
 UPDATE `creature_template` SET `ScriptName` = 'npc_hoo_aggro_stalker' WHERE `entry` = 40790;
-UPDATE `creature_template` SET `ScriptName` = 'npc_hoo_spatial_flux', `unit_flags` = 33554496 WHERE `entry` = 39612;
-UPDATE `creature_template` SET `ScriptName` = 'npc_hoo_energy_flux', `unit_flags` = 33554496, `speed_walk` = 5.5/2.5, `speed_run` = 5.5/2.5 WHERE `entry` = 44015;
+UPDATE `creature_template` SET `ScriptName` = 'npc_hoo_spatial_flux', `unit_flags` = 33554496 WHERE `entry` IN (39612, 48707;
```
missing )

Commit:
enum corrections

Commit:
sql fix and cosmetics

tkrokli commented on src/server/scripts/Kalimdor/HallsOfOrigination/halls_of_origination.cpp:
```diff
@@ -0,0 +1,421 @@
+            return spellEffectInfo && ValidateSpellInfo({ uint32(spellEffectInfo->BasePoints) });
+        }
+
+        void HandleScriptEffect(SpellEffIndex effIndex)
```
```cpp
src/server/scripts/Kalimdor/HallsOfOrigination/halls_of_origination.cpp:394:47:
 fatal error: unused parameter 'effIndex' [-Wunused-parameter]
        void HandleScriptEffect(SpellEffIndex effIndex)
                                              ^
```

tkrokli replied:
I hope the comment above was a simple enough.
All that is needed is to comment out the parameter.
(feel free to tell me if it should have been more detailed)

Mihapro replied:
Yeah it was. Aok told me to finish first part of sql to push it, ill do it now :P

Commit:
sql

Commit:
travis

Commit:
solve conflicts

Commit:
Misc

* Void Lord SAI
* Void Seeker cast spell on reset, condition for it
* spell effect correction, cosmetics...

Commit:
achievement hotfix

Aokromes commented:
@Mihapro https://app.codacy.com/app/TrinityCore/TrinityCore/pullRequest?prid=1497760

tkrokli commented:
Nice to see how Codacy works. To the point and similar to how real coders would comment on the PR content.

Although I fail to find issues 3 & 4 in the Codacy review, it looks to me as if there are only 2 issues listed ("Showing 2 files with new issues"). What am I overlooking/missing out on?

Aokromes commented:
I marked 2 false positives.

Commit:
Codacy issues

Commit:
catch up with master

Mihapro commented:
Due to the nature of students I have no time to continue scripting Rajh. Other content is 99% complete, with last one percent being the fact Arthur won't channel Hymn of Hope after my rewrite - teleport, orientate, root, channel. See [line 281](https://github.com/TrinityCore/TrinityCore/pull/20756/files#diff-a2e448d7fbda0a530b51930c723197b6R281) (EnterShieldPhase() and EVENT_CAST_SHIELD). Instantly interrupted by something? 🤔

Ovahlord commented:
U can port my Rajh script from 4.3.4.
And yes, Anhuurs channeling is getting interrupted by Shield of Light. Use a spell-linked-spell entry that will do it.
I am currently backporting the other implementations to 434 and noticed stability issues and non-effective code as well as wrong behaiviors that I am reworking atm.
One obvious crash is at isiset because one creature of the left side is not assigned to a formation and you skip many important nullptr checks in the instancescript

tkrokli commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_setesh.cpp:
```diff
@@ -0,0 +1,531 @@
+            float x = me->GetPositionX();
+            float y = me->GetPositionY();
+            float z = me->GetPositionZ();
+            Position dest = { x, y, me->GetMap()->GetWaterOrGroundLevel(me->GetPhases(), x, y, z) };
```
```cpp
src/server/scripts/Kalimdor/HallsOfOrigination/boss_setesh.cpp:343:77:
 fatal error: no member named 'GetPhases' in 'Creature'
            Position dest = { x, y, me->GetMap()->GetWaterOrGroundLevel(me->GetPhases(), x, y, z) };
                                                                        ~~  ^
1 error generated.
```

Ovahlord replied:
GetPhaseShift() since the phasing rewrite

tkrokli commented on src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp:
```diff
@@ -0,0 +1,858 @@
+
+    struct boss_isisetAI : public BossAI
+    {
+        boss_isisetAI(Creature* creature) : BossAI(creature, DATA_ISISET), _transitionPhase(false) { }
```
```
Member variable 'boss_isisetAI::_mirrorImageDied' is not initialized in the constructor.
```
https://app.codacy.com/app/TrinityCore/TrinityCore/pullRequest?prid=1497760
(... or is this a false positive?)

Aokromes committed:
Update boss_setesh.cpp

Commit:
init variable in constructor

Killyana commented:
@Mihapro can you add this to the PR: https://gitlab.com/trinitycore/TrinityCore_434/issues/176 ?

Commit:
Update SQL from 4.3.4

requested by @Killyana

jackpoz commented:
Closing this PR as it seems abandoned. Feel free to ask here to reopen it if anyone would like to work on it.
